packages/plugin-legacy/src/index.ts

```diff
@@ -331,24 +331,30 @@ function viteLegacyPlugin(options: Options = {}): Plugin[] {
         }
 
         return (chunkInfo) => {
-          let fileName =
+          const file =
             typeof fileNames === 'function' ? fileNames(chunkInfo) : fileNames
+          const fileArr = file.split('/')
+          let fileName = fileArr.pop()!
 
           if (fileName.includes('[name]')) {
             // [name]-[hash].[format] -> [name]-legacy-[hash].[format]
             fileName = fileName.replace('[name]', '[name]-legacy')
-          } else if (fileName.includes('[hash]')) {
-            // custom[hash].[format] -> [name]-legacy[hash].[format]
-            // custom-[hash].[format] -> [name]-legacy-[hash].[format]
-            // custom.[hash].[format] -> [name]-legacy.[hash].[format]
-            fileName = fileName.replace(/[.-]?\[hash\]/, '-legacy$&')
           } else {
-            // entry.js -> entry-legacy.js
-            // entry.min.js -> entry-legacy.min.js
-            fileName = fileName.replace(/(.+?)\.(.+)/, '$1-legacy.$2')
+            if (fileName.includes('[hash]') && !fileName.startsWith('[hash]')) {
+              // custom[hash].[format] -> [name]-legacy[hash].[format]
+              // custom-[hash].[format] -> [name]-legacy-[hash].[format]
+              // custom.[hash].[format] -> [name]-legacy.[hash].[format]
+              fileName = fileName.replace(/[.-]?\[hash\]/, '-legacy$&')
+            } else {
+              // entry.js -> entry-legacy.js
+              // entry.min.js -> entry-legacy.min.js
+              // [hash].[format] -> [hash]-legacy.[format]
+              // [hash]custom.[format] -> [hash]custom-legacy.[format]
+              fileName = fileName.replace(/(.+?)\.(.+)/, '$1-legacy.$2')
```

Comment:
I was proposing we use `legacy-[hash]` instead, but if we are trying to keep the legacy files as close as possible to their modern counterparts, then having `-legacy` at the end is better here.

Reply:
> if we are trying to keep the legacy files as close as possible to their modern counterparts, then having `-legacy` at the end is better here.

I think that doesn't happen because `[hash]` is different for modern chunks and legacy chunks.

Reply:
Ah... then `legacy-[hash]` would be better here, no? At least there is some kind of order, with different hashes then all legacy and modern chunks will end up intermixed.

Reply:
Yeah, I think `legacy-[hash]` would be better. I wonder if we should just use `[hash]` in that case though.

Reply:
If we do that, shouldn't we consider not adding `legacy` for every case when there is a `[hash]`? I think it is still useful to have `legacy` because you can't know what files are modern or legacy by looking only at the file name if not, no? (and the order will still be all mixed)

Reply:
If a user set it to `[hash]`, I guess the user doesn't want to give any information from the file name or wants to reduce the filename length. So in that case, the user wants the filename without `legacy`.

Reply:
If it is about not showing any info about the filename, then it doesn't hurt to have `legacy-`. The filename length is interesting. I wonder if we should let the users configure the names to override our defaults. I'm starting to think that we're crossing a limit on the complexity we are adding here, and maybe it would be better to have a clear rule for everything like adding `legacy-` to the basename in all cases (or maybe place all the legacy files in `legacy/` with the same name, the problem there may be that the file names could be confused when opening them in editors that only show the basename).

Reply:
If we are going further, I think it would be better to make users possible to pass the names like `[name][legacy:-]-[hash].[format]`. But for now, I think going with the current PR is fine. If someone requests a more fine-grained control, we can ask them to implement that.